### PR TITLE
[JENKINS-56397] Add HTTP session manager

### PR DESCRIPTION
**Pull request by jvz**
See [JENKINS-56397](https://issues.jenkins-ci.org/browse/JENKINS-56397).

### Proposed changelog entries

* Added HTTP session manager page for managing active Jenkins user sessions.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @daniel-beck @jeffret-b @Wadeck

### CODE REVIEW

File: core/src/main/java/jenkins/security/HttpSessionManager.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import static org.acegisecurity.context.HttpSessionContextIntegrationFilter.ACEGI_SECURITY_CONTEXT_KEY;
+
+@Restricted(NoExternalUse.class)
+@Extension
+public class HttpSessionManager extends ManagementLink implements SessionListener {
```
[the remark this answers is not in the document]
Comment (jvz) [reply core/src/main/java/jenkins/security/HttpSessionManager.java]: I don't see any classes in Jetty that have either name. Closest thing is [`SessionIdManager`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/SessionIdManager.java).

**Review by jeffret-b — COMMENTED** [review]
Some concerns:

The InvalidateAllSessions button isn't working right. It tries to download something to the client machine. It does seem to invalidate all sessions, though.

It would help to include a little descriptive text on the Jenkins management page to describe this capability. All / most of the other items there have a brief description.

I think "Terminate" would be a better word choice than "Invalidate".

It's kind of weird how when I click the "Invalidate" button, it gives a "Saved" message.

It looks like the localized messages aren't hooked up right. I see the key strings on the instance I built.

I don't think displaying the SessionId is useful and just clutters the view.

Is there a standard for displaying date-times in Jenkins? The format displayed is kind of messy.

Maybe multiple sessions from the same user should be grouped and managed together.  The Invalidate operation operates on all of them. Otherwise it is kind of confusing.

The UI could probably use some tweaking. Maybe we could see about getting an expert to provide some guidance.

I don't have an opinion as to whether this is a useful enhancement to Jenkins. It might be useful in some scenarios I suppose. It might be useful to try and raise the visibility of this enhancement to get more feedback as to its value. Maybe asking on the dev list?

**Comment by jvz** [comment]
> Some concerns:
> 
> The InvalidateAllSessions button isn't working right. It tries to download something to the client machine. It does seem to invalidate all sessions, though.

Possibly a bug from migrating the PR.

> It would help to include a little descriptive text on the Jenkins management page to describe this capability. All / most of the other items there have a brief description.

Agreed. Will add some help files and such.

> I think "Terminate" would be a better word choice than "Invalidate".

Makes sense.

> It's kind of weird how when I click the "Invalidate" button, it gives a "Saved" message.

That's a relic from using the FormApply thing. Maybe it's configurable?

> It looks like the localized messages aren't hooked up right. I see the key strings on the instance I built.

Most likely my bad. I put those in without testing it last minute.

> I don't think displaying the SessionId is useful and just clutters the view.

Seems like that's the case, yes. Should remove.

> Is there a standard for displaying date-times in Jenkins? The format displayed is kind of messy.

Not sure, but Functions might have something. A sort of fuzzy date (3 minutes ago) would be great.

> Maybe multiple sessions from the same user should be grouped and managed together. The Invalidate operation operates on all of them. Otherwise it is kind of confusing.

That would make sense! I like that idea.

> The UI could probably use some tweaking. Maybe we could see about getting an expert to provide some guidance.

Certainly.

> I don't have an opinion as to whether this is a useful enhancement to Jenkins. It might be useful in some scenarios I suppose. It might be useful to try and raise the visibility of this enhancement to get more feedback as to its value. Maybe asking on the dev list?

This would certainly need tweaking and improvements before being a useful feature. I wanted to submit this PoC from an earlier issue I had worked on but didn't end up needing.

**Comment by jvz** [comment]
> > Is there a standard for displaying date-times in Jenkins? The format displayed is kind of messy.
> 
> Not sure, but Functions might have something. A sort of fuzzy date (3 minutes ago) would be great.

Found it while looking through some code: `Util.getTimeSpanString()`.

File: core/src/main/resources/jenkins/security/HttpSessionManager/index.jelly [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                        </td>
+                    </tr>
+                </j:forEach>
+            </table>
+            <form method="post" action="invalidateAllSessions">
+                <f:submit value="${%InvalidateAllSessions}"/>
+            </form>
+        </l:main-panel>
+    </l:layout>
+</j:jelly>
```
Comment (jsoref) [inline core/src/main/resources/jenkins/security/HttpSessionManager/index.jelly]: Please include newlines at eof

**Comment by daniel-beck** [comment]
@jvz Do you intend to finish this PR? Or can someone else perhaps take over?

**Comment by jvz** [comment]
From what I remember, we ended up solving this problem through user seeds anyways, so I think this can be closed.

**Pull request closed without merging**